x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/metric/SecurityCacheMetrics.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.metric;
+
+import org.elasticsearch.common.cache.Cache;
+import org.elasticsearch.telemetry.metric.LongWithAttributes;
+import org.elasticsearch.telemetry.metric.MeterRegistry;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public final class SecurityCacheMetrics {
+
+    public static <K, V> List<AutoCloseable> registerAsyncCacheMetrics(MeterRegistry registry, Cache<K, V> cache, CacheType type) {
+        final List<AutoCloseable> metrics = new ArrayList<>();
+        metrics.add(
+            registry.registerLongAsyncCounter(
+                type.metricsPrefix + ".hit.total",
+                "Total number of cache hits.",
+                "count",
+                () -> new LongWithAttributes(cache.stats().getHits())
+            )
+        );
+        metrics.add(
+            registry.registerLongAsyncCounter(
+                type.metricsPrefix + ".miss.total",
+                "Total number of cache misses.",
+                "count",
+                () -> new LongWithAttributes(cache.stats().getMisses())
+            )
+        );
+        metrics.add(
+            registry.registerLongAsyncCounter(
+                type.metricsPrefix + ".eviction.total",
+                "Total number of cache evictions.",
+                "count",
+                () -> new LongWithAttributes(cache.stats().getEvictions())
+            )
+        );
+        metrics.add(
+            registry.registerLongGauge(
+                type.metricsPrefix + ".count.total",
+                "The current number of cache entries.",
+                "count",
+                () -> new LongWithAttributes(cache.count())
+            )
+        );
+        return metrics;
+    }
+
+    public enum CacheType {
+
+        API_KEY_AUTH_CACHE("es.security.api_key.auth_cache"),
```

Comment:
It's valid suggestion. I generally prefer to keep metrics aligned with setting names, but in this case I think that having `auth_cache`, `doc_cache` and `role_descriptor_cache`, makes better distinction between these. The `xpack.security.authc.api_key.cache` setting is generic and controls sizes  and TTLs (except for role descriptors) for all 3 caches. Under the hood, it is implemented as 3 caches. Also, in the future we may add a negative lookup caches (https://github.com/elastic/elasticsearch/issues/107369), so having prefix helps in avoiding confusions. Since you don't feel strongly, I will keep the current name.